streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,83 +78,123 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
+                                                   couldNotCreateTopic);
                     }
-                    streamsKafkaClient.createTopics(topicsToBeCreated, replicationFactor, windowChangeLogAdditionalRetention, metadata);
+                } catch (final InterruptedException fatalException) {
+                    final String error = "Could not create internal topics.";
+                    log.error(error, fatalException);
+                    throw new StreamsException(error, fatalException);
                 }
-                return;
-            } catch (StreamsException ex) {
-                log.warn("Could not create internal topics: {} Retry #{}", ex.getMessage(), i);
             }
-            // backoff
-            time.sleep(100L);
         }
-        throw new StreamsException("Could not create internal topics.");
+    }
+
+    private void verifyIfTopicExistsWithCorrectNumberOfPartitions(final String topic,
+                                                                  final InternalTopicConfig expectedNumberOfPartitions) {
+        final Map<String, Integer> existingTopicPartitions = getNumPartitions(Collections.singleton(topic));
+        if (validateTopicPartitions(Collections.singleton(expectedNumberOfPartitions), existingTopicPartitions).size() != 0) {
+            final String errorMsg = String.format("Could neither create topic %s nor receive any metadata about it.", topic);
+            log.error(errorMsg);
+            throw new StreamsException(errorMsg);
+        }
     }
 
     /**
      * Get the number of partitions for the given topics
      */
     public Map<String, Integer> getNumPartitions(final Set<String> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                existingTopicPartitions.keySet().retainAll(topics);
-
-                return existingTopicPartitions;
-            } catch (StreamsException ex) {
-                log.warn("Could not get number of partitions: {} Retry #{}", ex.getMessage(), i);
+        return getNumPartitions(topics, false);
+    }
+
+    private Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                  final boolean bestEffort) {
+        try {
+            final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+            final Map<String, KafkaFuture<TopicDescription>> futures = describeTopicsResult.values();
+
+            final Map<String, Integer> existingTopicPartitions = new HashMap<>();
+            for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+                try {
+                    final TopicDescription topicDescription = topicFuture.getValue().get();
```

Comment:
I see that we are less resilient :(

However, I don't understand why a long request timeout would resolve this? We are going to send a single request after the change and would still send a single request with a larger timeout (compared to sending up to 5 requests originally).

If AdminClient does not retry timeouts, we can only put a loop around this, and use the `AdminClient` retry config to retry manually until AdminClient is fixed.

Can we expect that `AdminClient` gets fixed before the next release? I would prefer to fix this issue on the right place instead of introducing workaround in our code base.

Reply:
I think there are two slight different cases that we are discussing here :)

First case is when the broker is unavailable, we do not yet send the request out even since we do not know who to send to with empty metadata, hence this request will sit in the admin client's queue until the broker comes back and the metadata gets refreshed;

Second case is after the request is sent, broker crashed, and even after it resumes the request is lost and admin client is doomed to throw timeout exception still (note if it is a broker soft failure like GC the broker can still send response back in time).

With a longer timeout the first case can be remedied, but not the second case. And I'd not expect `AdminClient` improve on this end before the next release. So maybe we should add a retry loop wrapping the `numPartitions` and `createTopics` call still.

Reply:
Questions:
 - for case (1), the timeout exception would still be returned via the future?
 - if yes, would the loop not cover both cases?
 - if not, what request timeout config would you suggest? (I still think, we don't need to change it because of the first two points that I think hold true)

Reply:
Case 1): yes, the queued requests will eventually be timed out with the exception set. I'm thinking the loop would cover it as it would catch `TimeoutException` and retry.